### PR TITLE
Update ansible-galaxy to handle role requirements

### PR DESCRIPTION
**Pull request by DaazKu**
##### SUMMARY

Implementation of https://github.com/ansible/proposals/issues/57

Contrarily to the role's meta definition "dependencies" it allows a roles to define "requirements/dependencies" that will **be only installed**.

- [x] Handle meta/requirements.yml
- [x] Integration tests
- [x] Changelog fragment
- [x] Documentation 

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

ansible-galaxy

##### ADDITIONAL INFO
There is no change what so ever in how requirement files are parsed and handled.

AFAIK the only potential problem of having this feature introduced in the ansible ecosystem is that it might cause problem with 3rd party tools that expect a `meta/requirement.ya?ml` that are not compatible with the current definition. I don't know what the ansible team philosophy is about these kind of things.

Assuming that this gets approved, I can also make a PR to update https://galaxy.ansible.com/docs/using/installing.html?highlight=dependencies#dependencies to move things forward a bit more quickly.

### CODE REVIEW

**Comment by DaazKu** [comment]
refresh unit tests

**Comment by DaazKu** [comment]
ready_for_review

Can't make the tests all pass. The  number of failures always match the number of errors and they fluctuate so I will assume that the test suite just has problem and that everything works as intended.

**Comment by DaazKu** [comment]
ready_for_review

**Comment by DaazKu** [comment]
T

**Comment by DaazKu** [comment]
Hi @mattclay

Sorry request your help but I can't for the love of god make the tests work on this PR.
I saw that you worked on test suite a bit so you might be able to help me here.

I'm now pretty damn sure that I did everything correctly:

- My new tests work (Tested locally and they pass on Shippable)
- I've rebased my PR today because I think some tests were broken prior to that
- The tests that fails are not the same. Most of the time it is one or multiple of the `aix/7.2/*` test(s) that fail and not necessarily the same.

As you can see I've been trying a couple time to have the test suite pass with no success.

I'd really like some kind of guidance on what to do here. IMO this is ready for review and would like the process to go forward.

Thanks

**Comment by mattclay** [comment]
@DaazKu I've restarted the failed tests. Some of the tests in the test suite are flakey and need to be restarted sometimes. Anyone with commit privileges on the repo is able to restart only the failed tests.

**Comment by DaazKu** [comment]
> @DaazKu I've restarted the failed tests. Some of the tests in the test suite are flakey and need to be restarted sometimes. Anyone with commit privileges on the repo is able to restart only the failed tests.

Thanks!

Couple of questions:
1. If I put this happens again can I still put the PR as ready for review or it won't be looking at because of the test suite failing?
2. If it is best to have the test suite succeed for a PR to be reviewed how to I determine the best person to ask for a restart of failing tests?

**Comment by mattclay** [comment]
@DaazKu Yes, you can still request reviews if the CI failures are unrelated to the changes in the PR. The best place to ask for re-running of failed jobs is in #ansible-devel in IRC on freenode.

**Comment by DaazKu** [comment]
ready_for_review

**Comment by DaazKu** [comment]
@samccann Anything else I need to do to get this reviewed?

**Comment by samccann** [comment]
@DaazKu  - I suggest you bring it to the core team meeting on IRC. There is one tomorrow on #ansible-meeting at 11AM ET. Add this PR to the agenda here -https://github.com/ansible/community/issues/532

File: changelogs/fragments/68288-galaxy-requirements-install-only.yml
Comment (mattclay) [inline changelogs/fragments/68288-galaxy-requirements-install-only.yml]: Typo in `ya?ml` ?

Comment (DaazKu) [reply changelogs/fragments/68288-galaxy-requirements-install-only.yml]: Nah it's "regex" like. But I can update to `yml` or `yaml` if it's clearer. Both are supported so it made sense to me.

File: lib/ansible/galaxy/role.py
Comment (bcoca) [inline lib/ansible/galaxy/role.py:384]: this reads both files and overwrites with the 2nd, i would just break the loop once the first is read successfully

File: lib/ansible/galaxy/role.py
Comment (bcoca) [inline lib/ansible/galaxy/role.py]: i would make this a warning and not return, the requirements would still be empty list and returned at the end, also this allows the loop to get to 2nd file and test if that works fine.

From comment above, if you process 1st file fine .. but then 2nd is error, you throw away the correct result, so combining both fixes shoudl get more reliable behaviour.

File: docs/docsite/rst/galaxy/user_guide.rst
Comment (bcoca) [inline docs/docsite/rst/galaxy/user_guide.rst:373]: version added tag

**Review by bcoca — CHANGES_REQUESTED** [review]

File: docs/docsite/rst/galaxy/user_guide.rst
Comment (DaazKu) [reply docs/docsite/rst/galaxy/user_guide.rst:373]: Already did a commit for that.

File: lib/ansible/galaxy/role.py
Comment (DaazKu) [reply lib/ansible/galaxy/role.py:384]: Valid point. I based myself off https://github.com/ansible/ansible/blob/43270332e7d10b297ff9d61cbeb129fcefa6f9db/lib/ansible/galaxy/role.py#L107-L118 though so it follows the same logic.

File: lib/ansible/galaxy/role.py
Comment (DaazKu) [reply lib/ansible/galaxy/role.py]: I based myself off existing as explained here https://github.com/ansible/ansible/pull/68288#discussion_r406295476
I think that breaking on the first file found would be less confusing since you should not have both a `yml` and `yaml` file.

File: lib/ansible/galaxy/role.py
Comment (DaazKu) [reply lib/ansible/galaxy/role.py:378]: I'm going to break on the first file found and keep throwing an exception if the parsing fails.

**Comment by DaazKu** [comment]
rebuild_failed

**Comment by DaazKu** [comment]
bot_broken

**Comment by DaazKu** [comment]
The `/rebuild_failed` commands appears to do nothing. If I'm wrong disregard.
https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#commands

**Comment by DaazKu** [comment]
ready_for_review

**Comment by DaazKu** [comment]
@bcoca Changes were mage, CI is passing.

File: changelogs/fragments/68288-galaxy-requirements-install-only.yml
Comment (samccann) [reply changelogs/fragments/68288-galaxy-requirements-install-only.yml]: my nickel would be to pick one (yml preferrably as that's what the doc uses).

File: docs/docsite/rst/galaxy/user_guide.rst
Comment (samccann) [inline docs/docsite/rst/galaxy/user_guide.rst:369]:
```suggestion

* using ``meta/requirements.yml``
* using ``meta/main.yml``

```

File: docs/docsite/rst/galaxy/user_guide.rst
Comment (samccann) [inline docs/docsite/rst/galaxy/user_guide.rst]:
```suggestion
Alternately, you can  specify role dependencies in the ``meta/main.yml`` file by providing a list of roles under the ``dependencies`` section. If the source of a role is Galaxy, you can simply specify the role in
```

File: test/integration/targets/ansible-galaxy/runme.sh
Comment (samccann) [inline test/integration/targets/ansible-galaxy/runme.sh]:
```suggestion
            # Prep git, because it doesn't work inside a docker container without it
```

File: docs/docsite/rst/galaxy/user_guide.rst
Comment (DaazKu) [reply docs/docsite/rst/galaxy/user_guide.rst]: Not a native english speaker but it seems that alternatively was a bit more what I was trying to convey:

> Alternately means several things taken in turn, consecutively, one after another. Alternately is an adverb formed by adding -ly to the word alternate, which comes from the Latin word alternatus meaning one after the other. Alternatively means another choice, on the other hand, another possibility.

**Comment by DaazKu** [comment]
ready_for_review

**Review by bcoca — APPROVED** [review]

File: docs/docsite/rst/galaxy/user_guide.rst
Comment (s-hertel) [inline docs/docsite/rst/galaxy/user_guide.rst]: This seems mostly redundant with the sentence above. Can you just add "to the ``roles_path``" to the end of the sentence above instead?

**Review by s-hertel — APPROVED** [review]

**Comment by DaazKu** [comment]
ready_for_review

**Review by samccann — APPROVED** [review]

**Comment by DaazKu** [comment]
LGTM?

**Comment by DaazKu** [comment]
@bcoca Can we merge this before CI goes stale or merge conflicts happen? Thanks.

**Pull request merged**
